elements/ip/ipfilter.cc

```diff
@@ -1208,12 +1208,40 @@ IPFilter::Parser::parse_test(int pos, bool negated)
 }
 
 void
+merge_programs(Classification::Wordwise::Program &p1, Classification::Wordwise::Program &p2)
+{
+    int step_offset = p1.ninsn();
+    int jump_target = step_offset;
+    if (p1.ninsn() == 0) {
+        // p1 program is empty.
+        // If it has a valid target, we always jump to it. p2 is ignored.
+        // If target is invalid, proceed as normal. Resulting tree == p2.
+        int target = -p1.output_everything();
+
+        //check if target is valid
+        if (target != Classification::j_never + 1) {
```

Comment:
It's been a few months since I wrote this, so please forgive me if I misunderstood some of the concepts for forget something!

Yes, so, this particular code is for when merging two programs/trees, p1 and p2, where p1 is empty.

The way rules are evaluated sequentially means P1 decisions override p2, so if P1 is empty and points to a valid rule 'target', then the whole of p2 becomes irrelevant. 

This check exits the function early (thus dropping p2) if p1 has a valid target, which I've defined to be "anything that's not an invalid target".

From reading the code and experimenting, it seemed that invalid targets have the numerical value of "j_never + 1". 

The Jumps enum in classification.hh, sets j_never to -2147483647, and so that +1 would be -2147483646. Which is also the value of the 'j_failure' enum.

It could be that I misunderstood this part, but in debugging that seemed to be the value the jumps were set to when they were "unlinked" or invalid.